### PR TITLE
`edge_update` example

### PR DESCRIPTION
**Pull request by Padarn**
PR that addresses discussion in https://github.com/pyg-team/pytorch_geometric/issues/813#issuecomment-962912361

### CODE REVIEW

**Comment by rusty1s** [comment]
Yes, I'm absolutely in favor of adding this. The interface looks good to me.

**Comment by Padarn** [comment]
Sorry no time during the week to work on this. I'll cleanup the code on the backend and push to this PR.

**Comment by Padarn** [comment]
I've added a basic implementation of this (obvious needs tests and documentation etc). 

Do you think it might be better to reuse the `propagate` function with some arguments e.g
```
x = self.propagate(...., edges=True)
```
Instead of splitting into the `self.edges_propagate`, I've added?

File: examples/edge_update.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+])
+path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', dataset)
+dataset = Planetoid(path, dataset, transform=transform)
+data = dataset[0]
+
+class EdgeMessagePassingExample(MessagePassing):
+
+    def forward(self, data: Data):
+        x = self.propagate(data.edge_index, edge_attr=data.edge_attr, x=data.x)
+        edge_attr = self.edges_propagate(data.edge_index, edge_attr=data.edge_attr, x=data.x)
```
Comment (rusty1s) [inline examples/edge_update.py:22]: We do not really propagate information here. I would prefer the earlier naming of `update_edge` (or something similar).

File: torch_geometric/nn/conv/message_passing.py [hunk 2]
```diff
@@ -115,8 +116,12 @@ def __init__(self, aggr: Optional[str] = "add",
         # Hooks:
         self._propagate_forward_pre_hooks = OrderedDict()
         self._propagate_forward_hooks = OrderedDict()
+        self._edges_propagate_forward_pre_hooks = OrderedDict()
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:119]: Do you think we need two hooks here? In contrast to `propagate`, edge updates only call a single function.

**Review by rusty1s — COMMENTED** [review]
I think this is great. I don't think we want to re-use `propagate` for this in order to prevent that our `propagate` functionality becomes too complicated (as it already is).

File: torch_geometric/nn/conv/message_passing.py (hunk 2 above)
Comment (Padarn) [reply torch_geometric/nn/conv/message_passing.py:119]: Good point, this seems unneeded, I will change to one.

File: examples/edge_update.py (hunk 1 above)
Comment (Padarn) [reply examples/edge_update.py:22]: I liked update_edge too, but I hate `update_edges` for the propagate and `edge_updates` for the message function.. which seemed a bit confusing.

**Comment by Padarn** [comment]
Hey @rusty1s - I've made some changes and add some basic tests based on the ones that were already there for message passing. What else would you like to see before we get something merged: Happy to add more tests or examples.

File: torch_geometric/nn/conv/message_passing.py [hunk 3]
```diff
@@ -353,6 +356,59 @@ def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
 
         return out
 
+    def edge_update(self, edge_index: Adj, size: Size = None, **kwargs):
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:359]: I think `size` can be dropped, as `size` is only useful for message aggregation.

File: torch_geometric/nn/conv/message_passing.py [hunk 4]
```diff
@@ -353,6 +356,59 @@ def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
...
+                for a in decomp_args
+            }
+            decomp_out = []
+
+        for i in range(decomposed_layers):
+            if decomposed_layers > 1:
+                for arg in decomp_args:
+                    kwargs[arg] = decomp_kwargs[arg][i]
+
+            coll_dict = self.__collect__(self.__user_args__, edge_index,
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:394]: I think this is problematic, as `self.__collect__` will collect all arguments (even for those in `propagate` although we only need the arguments passed to `edge_message`. I think we need to disentangle `__user_args__` in arguments for `propagate` and arguments for `edge_message`.

**Review by rusty1s — COMMENTED** [review]
Super, thanks for the update! Left some final comments before we can merge. One thing we need to add in a later PR is `torch.jit.script` support.

File: torch_geometric/nn/conv/message_passing.py (hunk 3 above)
Comment (Padarn) [reply torch_geometric/nn/conv/message_passing.py:359]: makes sense I'll update

**Comment by Padarn** [comment]
thanks @rusty1s - will update in the weekend (no time during the week usually)

File: torch_geometric/nn/conv/message_passing.py (hunk 4 above)
Comment (Padarn) [reply torch_geometric/nn/conv/message_passing.py:394]: understood, I will figure out how to clearly do this

**Comment by Padarn** [comment]
I've pushed changes for the discussed points.

But one thing I haven't wrapped my head around it how the jit stuff works, especially with the jinja templating: I'm a bit lost reading that code, any suggested reading to understand what the template is trying to achieve?

**Comment by Padarn** [comment]
Hey @rusty1s. Anything pending on this one? I'm happy to work on the JIT support too but may need some pointers in the right direction

**Comment by rusty1s** [comment]
Yes, let's merge this :) We can work on JIT support in a follow-up PR.

**Review by rusty1s — APPROVED** [review]

**Comment by Padarn** [comment]
Cool thanks! I'll study the JIT implementation we currently have and see if I can get the gist of it.

**Comment by rusty1s** [comment]
Yes, JIT is sadly troublesome to understand and to extend. If you feel like you are wasting too much time to understand, just let me know and I will try to integrate it by myself. Happy to answer any questions though :)

**Comment by Padarn** [comment]
started a PR here https://github.com/pyg-team/pytorch_geometric/pull/3765 - I think I have it mostly working but its a bit messy and there are a few minor problems.

**Pull request merged**